src/pkg/packager/filters/deploy.go

```diff
@@ -0,0 +1,194 @@
+					if groupSelected != nil {
+						return []types.ZarfComponent{}, fmt.Errorf(lang.PkgDeployErrMultipleComponentsSameGroup, groupSelected.Name, component.Name, component.DeprecatedGroup)
+					}
+
+					// Then append to the final list
+					selectedComponents = append(selectedComponents, component)
+					groupSelected = &component
+				}
+			}
+
+			// If nothing was selected from a group, handle the default
+			if groupSelected == nil && groupDefault != nil {
+				selectedComponents = append(selectedComponents, *groupDefault)
+			} else if len(groupedComponents[groupKey]) > 1 && groupSelected == nil && groupDefault == nil {
+				// If no default component was found, give up
+				componentNames := []string{}
+				for _, component := range groupedComponents[groupKey] {
+					componentNames = append(componentNames, component.Name)
+				}
+				return []types.ZarfComponent{}, fmt.Errorf(lang.PkgDeployErrNoDefaultOrSelection, strings.Join(componentNames, ", "))
+			}
+		}
+
+		// Check that we have matched against all requests
+		var err error
+		for _, requestedComponent := range f.requestedComponents {
+			if _, ok := matchedRequests[requestedComponent]; !ok {
+				closeEnough := []string{}
+				for _, c := range pkg.Components {
+					d := levenshtein.ComputeDistance(c.Name, requestedComponent)
+					if d <= 5 {
+						closeEnough = append(closeEnough, c.Name)
+					}
+				}
+				failure := fmt.Errorf(lang.PkgDeployErrNoCompatibleComponentsForSelection, requestedComponent, strings.Join(closeEnough, ", "))
+				if err != nil {
+					err = fmt.Errorf("%w, %w", err, failure)
+				} else {
+					err = failure
+				}
+			}
+		}
+		if err != nil {
+			return []types.ZarfComponent{}, err
+		}
+	} else {
+		for _, groupKey := range orderedComponentGroups {
+			group := groupedComponents[groupKey]
+			if len(group) > 1 {
+				if f.isInteractive {
+					component := interactive.SelectChoiceGroup(group)
+					selectedComponents = append(selectedComponents, component)
+				} else {
+					foundDefault := false
+					componentNames := []string{}
+					for _, component := range group {
+						// If the component is default, then use it
+						if component.Default {
+							selectedComponents = append(selectedComponents, component)
+							foundDefault = true
+							break
+						}
+						// Add each component name to the list
+						componentNames = append(componentNames, component.Name)
+					}
+					if !foundDefault {
+						// If no default component was found, give up
+						return []types.ZarfComponent{}, fmt.Errorf(lang.PkgDeployErrNoDefaultOrSelection, strings.Join(componentNames, ", "))
+					}
+				}
+			} else {
+				component := groupedComponents[groupKey][0]
+
+				if isRequired(component, useRequiredLogic) {
+					selectedComponents = append(selectedComponents, component)
+				} else if component.Default && !f.isInteractive {
+					selectedComponents = append(selectedComponents, component)
+				} else {
```

Comment:
It may be good to pull f.isInteractive() into a higher level if block - you would still fall into the else statement for optional components (it is just that that logic still knows about the non-interactive case) - this block from 155-189 can likely be simplified and feels a bit tacked on.